immscore/md5digest.h

```diff
@@ -40,35 +40,49 @@ class Md5Digest
         static unsigned char bin_buffer[128 / 8];
         static char hex_buf[34] = {'\0'};
         static char tag_buf[4] = {'\0'};
+        size_t nmemb_read;
+        long offset = OFFSET;
+        char *cur = hex_buf;
+        int res, err;
 
         FILE *fp = fopen(filename.c_str(), "r");
         if (!fp)
-            return "bad_checksum";
+            goto failed;
 
-        long offset = OFFSET;
-        fseek(fp, TAGOFFSET, SEEK_END);
-        size_t nmemb_read = fread(tag_buf, 4, 1, fp);
+        if (fseek(fp, TAGOFFSET, SEEK_END) == -1)
+            goto failed;
+
+        nmemb_read = fread(tag_buf, 4, 1, fp);
         if (nmemb_read != 1)
-            return "bad_checksum";
+            goto failed;
 
         if (!strncmp(tag_buf, "TAG", 3))
             offset += TAGOFFSET;
 
-        int res = fseek(fp, offset, SEEK_END);
-
+        res = fseek(fp, offset, SEEK_END);
         if (res)
             rewind(fp);
 
-        int err = md5_stream(fp, NUMBLOCKS, bin_buffer);
-        fclose(fp);
+        err = md5_stream(fp, NUMBLOCKS, bin_buffer);
         if (err)
-            return "bad_checksum";
+            goto failed;
+
+        err = fclose(fp);
+        if (err)
+            goto failed;
 
-        char *cur = hex_buf;
         for (int i = 0; i < (128 / 4 / 2); ++i)
             cur += sprintf(cur, "%02x", bin_buffer[i]);
 
         return hex_buf;
+
+failed:
+        LOG(ERROR) << "Error hashing file \"" << filename << "\": " << error_string(errno) << endl;
+        if (fp)
+            err = fclose(fp);
+        if (err)
+            goto failed;
```

Comment:
This could cause an infinite loop if fp != NULL and fails to close! If fclose fails here, I think you should just log it and exit.